Add a typecheck in the binary_rule method allowing a binary array as …

swifmaneum opened the pull request:
…rule; Adding an additional parameter powers_of_two to the binary_rule method allowing faster computation

lantunes commented on cellpylib/ca_functions.py:
```diff
@@ -84,13 +84,21 @@ def binary_rule(neighbourhood, rule, scheme=None):
+    else:
+        state_int = neighbourhood.dot(powers_of_two)
+    n = 2 ** len(neighbourhood)
+    if isinstance(rule, np.ndarray):
```
We should probably also accommodate python lists. Perhaps we can change this line to:
```
if isinstance(rule, (list, np.ndarray)):
```

lantunes commented on cellpylib/ca_functions.py:
```diff
@@ -84,13 +84,21 @@ def binary_rule(neighbourhood, rule, scheme=None):
+        state_int = neighbourhood.dot(powers_of_two)
+    n = 2 ** len(neighbourhood)
+    if isinstance(rule, np.ndarray):
+        rule_bin_array = rule
```
The `rule_bin_array` needs to be zero padded so that it has a minimum length, as is done in the `int_to_bits` function. For example:
```
radius = 1
rule = np.array([1, 1, 1, 1, 0])
rule_bin_array = np.pad(rule, ((2 ** (radius * 2 + 1)) - len(rule), 0), 'constant')
```
The code later on in the function expects the `rule_bin_array` to be of length 2^n, where n is the neighbourhood size (i.e. 2r+1). It slows things down to do this padding in the function, so it's best if it is done up front. This means that there should be a check that ensures the rule, when it's an array, has the right length. Alternatively, we can zero pad for the user if required.

lantunes commented on cellpylib/ca_functions.py:
```diff
@@ -84,13 +84,21 @@ def binary_rule(neighbourhood, rule, scheme=None):
+    :param rule: an int or a binary array indicating the cellular automaton rule number
     :param scheme: can be None (default) or 'nks'; if 'nks' is given, the rule numbering scheme used in NKS is used
+    :param powers_of_two: a pre-computed array containing the powers of two; can be None (default) or an array of length
+     len(neighbourhood); if an array is given, it will used to speed up the calculation of state_int
```
I notice that the powers of two, when provided, should be a list such as `[4,2,1]`. We should probably state this as an example in the docstring.

swifmaneum replied:
Added your example to the docstring

lantunes commented:
Here are a couple of test cases we can add (to `test_cellpylib.py`):
```
def test_binary_rule_nks(self):
    rule_number = 30
    radius = 1
    size = 149
    timesteps = 149

    expected = cpl.evolve(cpl.init_simple(size=size), timesteps=timesteps,
                          apply_rule=lambda n, c, t: cpl.binary_rule(n, rule_number, scheme="nks"), r=radius)

    powers_of_two = 2 ** np.arange(radius * 2 + 1)[::-1]
    rule = list(map(int, bin(rule_number)[2:]))
    rule_bin_array = np.pad(rule, ((2 ** (radius * 2 + 1)) - len(rule), 0), 'constant').tolist()
    actual = cpl.evolve(cpl.init_simple(size=size), timesteps=timesteps,
                        apply_rule=lambda n, c, t: cpl.binary_rule(n, rule_bin_array, scheme="nks", powers_of_two=powers_of_two),
                        r=radius)

    np.testing.assert_equal(expected.tolist(), actual.tolist())

def test_binary_rule_default(self):
    rule_number = 6667021275756174439087127638698866559
    radius = 3
    timesteps = 49
    init = np.array([[1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0,
            1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1]])

    expected = cpl.evolve(init, timesteps=timesteps,
                          apply_rule=lambda n, c, t: cpl.binary_rule(n, rule_number), r=radius)

    powers_of_two = 2 ** np.arange(radius * 2 + 1)[::-1]
    rule = list(map(int, bin(rule_number)[2:]))
    rule_bin_array = np.pad(rule, ((2 ** (radius * 2 + 1)) - len(rule), 0), 'constant')
    actual = cpl.evolve(init, timesteps=timesteps,
                        apply_rule=lambda n, c, t: cpl.binary_rule(n, rule_bin_array, powers_of_two=powers_of_two),
                        r=radius)

    np.testing.assert_equal(expected.tolist(), actual.tolist())
```

I also changed your implementation of `binary_rule` locally, on line 98, before running the tests above:
```
if isinstance(rule, (list, np.ndarray)):
    assert len(rule) == n
    rule_bin_array = rule
```

swifmaneum commented:
I added your suggestions. Right now there's only the `assert len(rule) == n` for checking that the rule array is sized correctly but I agree that we could add a check and pad the rule array if necessary. I also added an additional assert ensuring that `powers_of_two` has the right size.

lantunes commented:
> I added your suggestions. Right now there's only the `assert len(rule) == n` for checking that the rule array is sized correctly but I agree that we could add a check and pad the rule array if necessary. I also added an additional assert ensuring that `powers_of_two` has the right size.

Sounds good, thank you for contributing this!

lantunes reviewed (approved):
Looks good to me.

lantunes commented:
closes #3
